muxers/mplex/src/lib.rs

```diff
@@ -46,30 +48,38 @@ impl UpgradeInfo for MplexConfig {
 
 impl<C> InboundUpgrade<C> for MplexConfig
 where
-    C: AsyncRead + AsyncWrite + Unpin,
+    C: AsyncRead + AsyncWrite + Unpin + Send + 'static,
 {
-    type Output = Multiplex<C>;
+    type Output = Multiplex;
     type Error = io::Error;
     type Future = future::Ready<Result<Self::Output, io::Error>>;
 
     fn upgrade_inbound(self, socket: C, _: Self::Info) -> Self::Future {
         future::ready(Ok(Multiplex {
-            io: Mutex::new(io::Multiplexed::new(socket, self)),
+            io: Arc::new(Mutex::new(io::Multiplexed::new(
+                SubstreamBox::new(socket),
+                self,
+            ))),
+            outbound_substreams: OutboundSubstreams::default(),
         }))
     }
 }
 
 impl<C> OutboundUpgrade<C> for MplexConfig
 where
-    C: AsyncRead + AsyncWrite + Unpin,
+    C: AsyncRead + AsyncWrite + Unpin + Send + 'static,
 {
-    type Output = Multiplex<C>;
+    type Output = Multiplex;
     type Error = io::Error;
     type Future = future::Ready<Result<Self::Output, io::Error>>;
 
     fn upgrade_outbound(self, socket: C, _: Self::Info) -> Self::Future {
         future::ready(Ok(Multiplex {
-            io: Mutex::new(io::Multiplexed::new(socket, self)),
+            io: Arc::new(Mutex::new(io::Multiplexed::new(
+                SubstreamBox::new(socket),
```

Comment:
@elenaf9 I remember why I didn't name it `SubstreamBox` initially! Because it is also used here to erase the type of the socket which would otherwise also show up in the type signature of `libp2p_mplex::Substream`.

However, I think that is actually not a problem because _now_ we box the substream later on anyway. In an earlier design we didn't!